### DISCUSSION TITLE
[Feature] Given a photo, find similar photos

### DISCUSSION THREAD

**Discussion by simone-viozzi** [immich-app/immich#4509] [Feature Request]
### The feature

I propose a feature that uses the existing clip embeddings to query similar photos.

## proposed user interface

From an image, add an option under the 3 dot menu'
![image](https://github.com/immich-app/immich/assets/19674168/5ddccb86-cb15-4ee5-9386-39472917b8e4)

We can reuse the same interface as a text query
![image](https://github.com/immich-app/immich/assets/19674168/41919983-6eb1-4a08-9ee8-62a55ac514e1)

## implementation

It should be possible to use the embeddings of an image, to query the clip embeddings of all the other images. 
Same that happen with a text query, but using the ones from one image.

### Platform

- [X] Server
- [X] Web
- [X] Mobile

**Comment by nodis** [comment]
This feature is very important and worth looking forward to

**Comment by akropp** [comment]
Would definitely be useful!

**Comment by Rahulsharma0810** [comment]
+1

**Comment by cunicode** [comment]
reference: Librephots has this as a default on the info-panel for each photo 
[https://github.com/search?q=repo%3ALibrePhotos%2Flibrephotos%20similar&type=code](https://github.com/search?q=repo%3ALibrePhotos%2Flibrephotos%20similar&type=code)

**Comment by yznts** [comment]
Feature would be really useful for cleaning up library from lots of similar photos (f.e multiple shots from the same angle)

**Comment by ThreeAurora** [comment]
This feature is very important.

Right now it is not possible to find similar images or to locate the position of the searched image in the album.

This seems to make the search results "very independent and difficult to manipulate".

**Comment by axgd-code** [comment]
The feature will be extremely interesting for creating albums and would also provide an advantage over Google Photos

**Reply by ThreeAurora** [reply]
Yes, but it's been a year and there has been no reply

**Reply by klejejs** [reply]
There is a similar photo detection feature in works, in fact some its code has already been merged, so I think we can expect something in upcoming releases.

**Reply by NicholasFlamy** [reply]
#8228 is the backend.

**Reply by quinnaissance** [reply]
> #8228 is the backend.

So as far as I understand it, the only/main backend difference here to finding duplicates would be adjusting the threshold?

**Comment by aviv926** [comment]
With the new feature that finds duplicates could this be easy to implement? It will be interesting to know what the team thinks about it. Can you put it on Discord?

**Reply by bo0tzz** [reply]
@mertalev was already playing with it iirc

**Reply by aviv926** [reply]
> @mertalev was already playing with it iirc

Nicee we will wait for the magician 🪄😁

**Reply by mertalev** [reply]
Nope, I didn't work on this. I think @NicholasFlamy was interested in this feature and was waiting for duplicate detection to be added.

**Reply by NicholasFlamy** [reply]
@johnjamespj wanted to work on this and I agreed to help but I don't think either of us have done anything yet (I know I haven't).

**Reply by SelfSufficient08** [reply]
This would be a great feature to add.  I often search for a photo, find it and then want to find photos similiar or even find photos from that same day or timeframe.

**Comment by Broyojo** [comment]
It would also be very important to make a "find similar faces" functionality that measures similarity between provided image of a face or image with a face in it and find other images with similar faces in them.

**Reply by NicholasFlamy** [reply]
That already exists, you can tap the detected person in the photo and it will show all the photos with that face.

![Screenshot_20241010_135005_Brave](https://github.com/user-attachments/assets/f4bcc889-8778-4671-b052-14712f6f10dc)
![20241010_135424](https://github.com/user-attachments/assets/c7e473e9-435e-4125-97af-33964b73b04a)

**Reply by Broyojo** [reply]
Yes, however this requires the face to be in an existing photo in the database and needs to be a valid face (exists across several photos, etc.). I want to be able to give immich and image of a face and it can find me images with the most similar faces

**Comment by jerji** [comment]
At this point I have almost 100k originals and this feature would be awesome for finding sets and duplicates more specifically.

**Comment by gloryandel** [comment]
This is amazing for viewing and exploring similar photos.

**Reply by MariasStory** [reply]
Should be really easy to implement using searchRepository.searchSmart:
https://github.com/immich-app/immich/blob/dd9feeec45d272dbe2c2100e40e0f465d960a964/server/src/services/search.service.ts#L95
Basically, knowing the embedding of the image search similar in pgvector database.
I don't understand why it is not implemented.

**Reply by mertalev** [reply]
The way I imagine this feature is that you can scroll down from the image and have it show related content (similar photos, albums this photo is in, etc.). I'm not much of a frontend dev, though, and the focus of the dev team is elsewhere. PRs are welcome if anyone wants to take a stab at it.

**Reply by cunicode** [reply]
This repo does it it nicely:  https://github.com/sankalp1999/Embeddit  

search -> shows results -> tap an image -> navigate across results -> use any given image as a "search by image"

**Comment by PathToLife** [comment]
I have a draft PR demoing the feature but still needs a few changes.

Can speed things up if there's some activity. Main concern is how the similar search bar UI would look.

- When you open a photo for viewing a similar search button is added to the top right corner.
   - No drop down menu at this time as I think similar search will be used quit a bit for sets if photos that aren't stacked
   - This can be adjusted to be hidden in the dropdown if requested
- Clicking on the similar search button will take you to a prefilled search results page.
- Middle clicking will open this search in a new tab

https://github.com/immich-app/immich/pull/17061
